apps/src/templates/progress/progressHelpers.js

```diff
@@ -180,6 +241,176 @@ export const processedLevel = level => {
     levelNumber: level.kind === LevelKind.unplugged ? undefined : level.title,
     isConceptLevel: level.is_concept_level,
     bonus: level.bonus,
-    sublevels: level.sublevels
+    sublevels:
+      level.sublevels &&
+      level.sublevels.map(level => processedLevel(level, true))
+  };
+};
+
+/**
+ * Given a level progress object that we get from the server using either
+ * /api/user_progress or /dashboardapi/section_level_progress,
+ * extracts the result, appropriately discerning a locked/submitted
+ * result for certain levels.
+ */
+const getLevelResult = serverProgress => {
+  if (serverProgress.status === LevelStatus.locked) {
+    return TestResults.LOCKED_RESULT;
+  }
+  if (serverProgress.readonly_answers) {
+    return TestResults.READONLY_SUBMISSION_RESULT;
+  }
+  if (serverProgress.submitted) {
+    return TestResults.SUBMITTED_RESULT;
+  }
+
+  return serverProgress.result || TestResults.NO_TESTS_RUN;
+};
+
+export const levelProgressFromServer = serverObject => {
```

Comment:
For this method and other "FromServer" methods below, it might useful to add a comment indicating where to find the corresponding code on the server side.